xarray/testing/strategies.py

```diff
@@ -0,0 +1,761 @@
+        max_size=max_ndims,
+        unique=True,
+    )
+
+
+def dimension_sizes(
+    min_ndims: int = 0,
+    max_ndims: int = 3,
+    min_length: int = 1,
+    max_length: int = None,
+) -> st.SearchStrategy[Mapping[str, int]]:
+    """
+    Generates an arbitrary mapping from dimension names to lengths.
+
+    Requires the hypothesis package to be installed.
+
+    Parameters
+    ----------
+    min_ndims: int, optional
+        Minimum number of dimensions in generated list.
+        Default is 1.
+    max_ndims: int, optional
+        Maximum number of dimensions in generated list.
+        Default is 3.
+    min_length: int, optional
+        Minimum size of a dimension.
+        Default is 1.
+    max_length: int, optional
+        Minimum size of a dimension.
+        Default is `min_length` + 5.
+    """
+
+    if max_length is None:
+        max_length = min_length + 5
+
+    return st.dictionaries(
+        keys=names,
+        values=st.integers(min_value=min_length, max_value=max_length),
+        min_size=min_ndims,
+        max_size=max_ndims,
+    )
+
+
+# Is there a way to do this in general?
+# Could make a Protocol...
+T_Array = Any
+
+
+@st.composite
+def variables(
+    draw: st.DrawFn,
+    data: st.SearchStrategy[T_Array] = None,
+    dims: Union[
+        st.SearchStrategy[List[str]], st.SearchStrategy[Mapping[str, int]]
+    ] = None,
+    attrs: st.SearchStrategy[Mapping] = None,
+    convert: Callable[[np.ndarray], T_Array] = lambda a: a,
+) -> st.SearchStrategy[xr.Variable]:
+    """
+    Generates arbitrary xarray.Variable objects.
+
+    Follows the signature of the xarray.Variable constructor, but you can also pass alternative strategies to generate
+    either numpy-like array data or dimension names.
+
+    Passing nothing will generate a completely arbitrary Variable (backed by a numpy array).
+
+    Requires the hypothesis package to be installed.
+
+    Parameters
+    ----------
+    data: Strategy generating array-likes, optional
+        Default is to generate numpy data of arbitrary shape, values and dtype.
+    dims: Strategy for generating the dimensions, optional
+        Can either be a strategy for generating a list of string dimension names,
+        or a strategy for generating a mapping of string dimension names to integer lengths along each dimension.
+        If provided in the former form the lengths of the returned Variable will either be determined from the
+        data argument if given or arbitrarily generated if not.
+        Default is to generate arbitrary dimension names for each axis in data.
+    attrs: Strategy which generates dicts, optional
+    convert: Callable
+        Function which accepts one numpy array and returns one numpy-like array of the same shape.
+        Applied to the data after it is drawn from the `data` strategy provided.
+        Useful for converting numpy arrays to other types of arrays, e.g. sparse arrays.
+        Default is a no-op.
```

Comment:
Given that we have https://hypothesis.readthedocs.io/en/latest/numpy.html#array-api, why not just allow the `data` strategy to provide whatever type of array the user wants?  

The absolute worst case is that they have to write `data=xs.map(f)` instead of `data=xs, convert=f`, and that's *shorter*.  I'm also strongly in favor of providing this kind of "smooth ramp" between helper functions and code-it-yourself style, because almost all users are going to need to go off the beaten path of provided APIs sometimes.  It's best for everyone if the transition is smooth so that's not too much of a shock.

Reply:
The motivation was the way @keewis designed the subclassable test suite for duck array types in #4972 , where creating any numpy-like type is handled by one `create_data` function.

> The absolute worst case is that they have to write data=xs.map(f)

This is a very good point though, we should probably just recommend this instead and document it clearly.

Reply:
Removed the `convert` parameter in favour of supplying a different strategy to `data`.